File: doordeck-sdk/src/mingwMain/resources/csharp/Wrapper/ContextManager.cs
using System.Security.Cryptography.X509Certificates;
using Doordeck.Headless.Sdk.Model;
using Doordeck.Headless.Sdk.Utilities;

namespace Doordeck.Headless.Sdk.Wrapper;

/// <summary>
/// Context operations complete against local state, so the SDK invokes their callback before the
/// native call returns and <see cref="Dispatcher.CallSync{T}"/> never actually waits. The two that
/// reach the server stay asynchronous.
/// </summary>
public class ContextManager
{
    public ApiEnvironment GetApiEnvironment() =>
        Enum.Parse<ApiEnvironment>(Dispatcher.CallSync<string>("context.getApiEnvironment"));

    public void SetApiEnvironment(ApiEnvironment apiEnvironment) =>
        Dispatcher.CallSync<object>("context.setApiEnvironment", apiEnvironment.ToString());

    public void SetCloudAuthToken(string token) =>
        Dispatcher.CallSync<object>("context.setCloudAuthToken", token);

    public string GetCloudAuthToken() =>
        Dispatcher.CallSync<string>("context.getCloudAuthToken");

    public Task<bool> IsCloudAuthTokenInvalidOrExpired(bool checkServerInvalidation) =>
        Dispatcher.Call<bool>("context.isCloudAuthTokenInvalidOrExpired", checkServerInvalidation.ToString());

    public void SetCloudRefreshToken(string token) =>
        Dispatcher.CallSync<object>("context.setCloudRefreshToken", token);

    public string GetCloudRefreshToken() =>
        Dispatcher.CallSync<string>("context.getCloudRefreshToken");

    public void SetFusionHost(string host) =>
        Dispatcher.CallSync<object>("context.setFusionHost", host);

    public string GetFusionHost() =>
        Dispatcher.CallSync<string>("context.getFusionHost");

    public void SetFusionAuthToken(string token) =>
        Dispatcher.CallSync<object>("context.setFusionAuthToken", token);

    public string GetFusionAuthToken() =>
        Dispatcher.CallSync<string>("context.getFusionAuthToken");

    public void SetUserId(Guid userId) =>
        Dispatcher.CallSync<object>("context.setUserId", userId.ToString());

    public Guid GetUserId() =>
        Guid.Parse(Dispatcher.CallSync<string>("context.getUserId"));

    public void SetUserEmail(string email) =>
        Dispatcher.CallSync<object>("context.setUserEmail", email);

    public string GetUserEmail() =>
        Dispatcher.CallSync<string>("context.getUserEmail");

    public void SetCertificateChain(List<X509Certificate> certificateChain) =>
        Dispatcher.CallSync<object>("context.setCertificateChain", certificateChain.CertificateChainToString());

    public List<X509Certificate> GetCertificateChain() =>
        Dispatcher.CallSync<string>("context.getCertificateChain").StringToCertificateChain();

    public bool IsCertificateChainInvalidOrExpired() =>
        Dispatcher.CallSync<bool>("context.isCertificateChainInvalidOrExpired");

    public void SetKeyPair(byte[] publicKey, byte[] privateKey) =>
        Dispatcher.CallSync<object>("context.setKeyPair", new
        {
            publicKey = publicKey.EncodeByteArrayToBase64(),
            privateKey = privateKey.EncodeByteArrayToBase64()
        });

    public KeyPair GetKeyPair() =>
        Utils.FromJson<KeyPair>(Dispatcher.CallSync<string>("context.getKeyPair"));

    public void SetKeyPairVerified(byte[]? publicKey) =>
        Dispatcher.CallSync<object>("context.setKeyPairVerified", publicKey?.EncodeByteArrayToBase64());

    public bool IsKeyPairVerified() =>
        Dispatcher.CallSync<bool>("context.isKeyPairVerified");

    public bool IsKeyPairValid() =>
        Dispatcher.CallSync<bool>("context.isKeyPairValid");

    public void SetOperationContext(Guid userId, List<X509Certificate> certificateChain, byte[] publicKey,
        byte[] privateKey, bool isKeyPairVerified) =>
        Dispatcher.CallSync<object>("context.setOperationContext", new
        {
            userId,
            certificateChain = certificateChain.CertificateChainToString(),
            publicKey,
            privateKey,
            isKeyPairVerified
        });

    public Task<ContextState> GetContextState(bool checkServerInvalidation) =>
        Dispatcher.Call<ContextState>("context.getContextState", checkServerInvalidation.ToString());

    public void ClearContext() =>
        Dispatcher.CallSync<object>("context.clearContext");
}

File: doordeck-sdk/src/commonMain/kotlin/com/doordeck/multiplatform/sdk/context/ContextManager.kt
package com.doordeck.multiplatform.sdk.context

import com.doordeck.multiplatform.sdk.model.data.ApiEnvironment
import com.doordeck.multiplatform.sdk.model.data.Crypto
import kotlin.js.JsExport
import kotlin.native.CName

@JsExport
interface ContextManager {

    /**
     * Sets the API environment on which the SDK will operate, the provided value will be automatically stored in secure storage.
     *
     * It can be changed anytime, although it is recommended to set it up through the `SdkConfig` builder.
     */
    fun setApiEnvironment(apiEnvironment: ApiEnvironment)

    /**
     * Retrieves the current API environment.
     */
    @CName("getApiEnvironment")
    fun getApiEnvironment(): ApiEnvironment

    /**
     * Sets the cloud authentication token, the provided value will be automatically stored in secure storage.
     *
     * It can be changed anytime, although it is recommended to set it up through the `SdkConfig` builder.
     */
    @CName("setCloudAuthToken")
    fun setCloudAuthToken(token: String)

    /**
     * Retrieves the cloud authentication token.
     */
    @CName("getCloudAuthToken")
    fun getCloudAuthToken(): String?

    /**
     * Checks whether the cloud authentication token is invalid (e.g., null, malformed) or expired.
     * (we consider it expired if it will expire within the next [com.doordeck.multiplatform.sdk.util.MIN_TOKEN_LIFETIME_DAYS] days).
     */
    @CName("isCloudAuthTokenInvalidOrExpired")
    fun isCloudAuthTokenInvalidOrExpired(): Boolean

    /**
     * Sets the cloud refresh token, the provided value will be automatically stored in secure storage.
     *
     * It can be changed anytime, although it is recommended to set it up through the `SdkConfig` builder.
     */
    @CName("setCloudRefreshToken")
    fun setCloudRefreshToken(token: String)

    /**
     * Retrieves the cloud refresh token.
     */
    @CName("getCloudRefreshToken")
    fun getCloudRefreshToken(): String?

    /**
     * Sets the fusion host, the provided value will be automatically stored in secure storage.
     *
     * It can be changed anytime, although it is recommended to set it up through the SdkConfig builder.
     */
    @CName("setFusionHost")
    fun setFusionHost(host: String)

    /**
     * Retrieves the fusion host.
     */
    @CName("getFusionHost")
    fun getFusionHost(): String

    /**
     * Sets the Fusion authentication token, the provided value will be automatically stored in secure storage.
     */
    @CName("setFusionAuthToken")
    fun setFusionAuthToken(token: String)

    /**
     * Retrieves the Fusion authentication token.
     */
    @CName("getFusionAuthToken")
    fun getFusionAuthToken(): String?

    /**
     * Sets the user ID associated with the context, the provided value will be automatically stored in secure storage.
     */
    @CName("setUserId")
    fun setUserId(userId: String)

    /**
     * Retrieves the user identifier.
     */
    @CName("getUserId")
    fun getUserId(): String?

    /**
     * Sets the user email associated with the context, the provided value will be automatically stored in secure storage.
     */
    @CName("setUserEmail")
    fun setUserEmail(email: String)

    /**
     * Retrieves the user email.
     */
    @CName("getUserEmail")
    fun getUserEmail(): String?

    /**
     * Sets the certificate chain for the context, the provided value will be automatically stored in secure storage.
     */
    fun setCertificateChain(certificateChain: List<String>)

    /**
     * Retrieves the certificate chain.
     */
    fun getCertificateChain(): List<String>?

    /**
     * Checks whether the certificate chain is invalid (e.g., null, malformed) or expired.
     * (we consider it expired if it will expire within the next [com.doordeck.multiplatform.sdk.crypto.MIN_CERTIFICATE_LIFETIME_DAYS] days).
     */
    @CName("isCertificateChainInvalidOrExpired")
    fun isCertificateChainInvalidOrExpired(): Boolean

    /**
     * Sets the key pair for the context, the provided values will be automatically stored in secure storage.
     */
    fun setKeyPair(publicKey: ByteArray, privateKey: ByteArray)

    /**
     * Retrieves the key pair.
     */
    fun getKeyPair(): Crypto.KeyPair?

    /**
     * Sets the key pair verification status, the provided values will be automatically stored in secure storage.
     */
    fun setKeyPairVerified(publicKey: ByteArray?)

    /**
     * Retrieves the key pair verification status.
     */
    @CName("isKeyPairVerified")
    fun isKeyPairVerified(): Boolean

    /**
     * Checks whether the current key pair is not null and valid by signing a small piece of text and verifying it.
     */
    @CName("isKeyPairValid")
    fun isKeyPairValid(): Boolean

    /**
     * Sets all necessary fields to perform secure operations, the provided values will be automatically stored in secure storage.
     */
    fun setOperationContext(userId: String, certificateChain: List<String>, publicKey: ByteArray, privateKey: ByteArray, isKeyPairVerified: Boolean = true)

    /**
     * Sets all necessary fields to perform secure operations in JSON format, the provided values will be automatically stored in secure storage.
     */
    @CName("setOperationContextJson")
    fun setOperationContextJson(data: String)

    /**
     * Clears all the values stored in secure storage.
     */
    @CName("clearContext")
    fun clearContext()
}